From 03485312488c8ca65b08f2bd44ab00383a599cea Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Tue, 26 Jun 2018 09:39:27 +0100
Subject: [PATCH] Hide AVX-512 from guests by default

It can be enabled by providing "cpuid=avx512" in Xen's cmdline.

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/arch/x86/cpuid.c b/xen/arch/x86/cpuid.c
index e3aaf842a0..9417092bbc 100644
--- a/xen/arch/x86/cpuid.c
+++ b/xen/arch/x86/cpuid.c
@@ -19,6 +19,8 @@ static const uint32_t hvm_shadow_featuremask[] = INIT_HVM_SHADOW_FEATURES;
 static const uint32_t hvm_hap_featuremask[] = INIT_HVM_HAP_FEATURES;
 static const uint32_t deep_features[] = INIT_DEEP_FEATURES;
 
+bool __initdata opt_avx512 = false;
+
 static int __init parse_xen_cpuid(const char *s)
 {
     const char *ss;
@@ -36,6 +38,12 @@ static int __init parse_xen_cpuid(const char *s)
         if ( !ss )
             ss = strchr(s, '\0');
 
+        if ( (val = parse_boolean("avx512", s, ss)) >= 0 )
+        {
+            opt_avx512 = val;
+            goto next;
+        }
+
         /* Skip the 'no-' prefix for name comparisons. */
         feat = s;
         if ( strncmp(s, "no-", 3) == 0 )
@@ -80,6 +88,7 @@ static int __init parse_xen_cpuid(const char *s)
         if ( mid )
             rc = -EINVAL;
 
+    next:
         s = ss + 1;
     } while ( *ss );
 
diff --git a/xen/arch/x86/setup.c b/xen/arch/x86/setup.c
index 27981adc0b..da6e225728 100644
--- a/xen/arch/x86/setup.c
+++ b/xen/arch/x86/setup.c
@@ -1623,6 +1623,9 @@ void __init noreturn __start_xen(unsigned long mbi_p)
     if ( opt_invpcid && cpu_has_invpcid )
         use_invpcid = true;
 
+    if ( !opt_avx512 )
+        setup_clear_cpu_cap(X86_FEATURE_AVX512F);
+
     init_speculation_mitigations();
 
     init_idle_domain();
diff --git a/xen/include/asm-x86/cpuid.h b/xen/include/asm-x86/cpuid.h
index 6c0c403788..a7ca8139f5 100644
--- a/xen/include/asm-x86/cpuid.h
+++ b/xen/include/asm-x86/cpuid.h
@@ -13,6 +13,8 @@
 
 #include <public/sysctl.h>
 
+extern bool opt_avx512;
+
 extern const uint32_t known_features[FSCAPINTS];
 extern const uint32_t special_features[FSCAPINTS];
 
